File: Topk.java
// Time Complexity : O(N+MlogK)   
// Space Complexity : O(k)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no

// Approach

// We can solve this using map and priority queue
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        if(nums == null || nums.length == 0)return new int[]{};
        int[] result = new int[k];
        HashMap<Integer,Integer> map = new HashMap<>();
        PriorityQueue<int[]> pq = new PriorityQueue<>((a,b)-> a[1] - b[1]);
    
        for(int n : nums){
            map.put(n, map.getOrDefault(n ,0)+1);
        }
        for(int key : map.keySet()){
            pq.add(new int[]{key, map.get(key)});
            if(pq.size() >k){
                pq.poll();
            }
        }
        for(int j = k-1; j>=0; j--){
            int[] polled = pq.poll();
            result[j] = polled[0];
        }
        return result;
    }
}

File: topk.py
# Time: O(n)
# Space: O(n)
# Did it run on Leetcode: yes

from collections import defaultdict
from Queue import PriorityQueue

class Solution(object):
    def topKFrequent(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: List[int]
        """
        # O(nlogk)
        hmap = defaultdict(int)
        for num in nums:
            hmap[num] += 1

        pq = PriorityQueue()
        for num, freq in hmap.items():
            pq.put((freq, num))
            if pq.qsize() > k:
                pq.get()

        res = []
        while not pq.empty():
            res.append(pq.get()[1])

        return res[::-1]
# using bucketsort
# O(n)
        # hmap = defaultdict(int)
        # for num in nums:
        #     hmap[num] += 1
        # freqmap={}
        # minel=float('inf')
        # maxel=float('-inf')
        # for key in hmap:
        #     freq=hmap[key]
        #     maxel=max(maxel,freq)
        #     minel=min(minel, freq)
        #     if freq not in freqmap:
        #         freqmap[freq]=[]
        #     freqmap[freq].append(key)
        # res=[]
        
        # for j in range(maxel, minel+1, -1):
        #     if j in freqmap:
        #         li=freqmap[j]
        #         if(len(li)==0):
        #             continue
        #         for i in range(len(li)):
        #             if(k>0):
        #                 res.append(li[i])
        #                 k-=1
        #             else:
        #                 break
        # return res
